File: dart/commection/test/example_test.dart
import 'package:commection/example.dart';
import 'package:test/test.dart';

void main() {
  final implement = Implement(hello: () async {
    return 'ok!!!';
  });

  test('text is empty', () async {
    expect(await TextIsEmpty(TextLiteral('')).evaluate(implement), true);
  });

  test('text join', () async {
    expect(
      await TextJoin(TextLiteral('sample: '), Hello()).evaluate(implement),
      'sample: ok!!!',
    );
  });

  test('if true', () async {
    expect(
      await If(
        condition: TextIsEmpty(TextLiteral('')),
        thenExpr: TextLiteral('T'),
        elseExpr: TextLiteral('F'),
      ).evaluate(implement),
      'T',
    );
  });

  test('if false', () async {
    expect(
      await If(
        condition: TextIsEmpty(TextLiteral('aa')),
        thenExpr: TextLiteral('T'),
        elseExpr: TextLiteral('F'),
      ).evaluate(implement),
      'F',
    );
  });

  test('optional match', () async {
    expect(
      await OptionalMatch(
        optional: TextLiteral('aa'),
        someExpr: (value) => TextJoin(TextLiteral('Some('), value),
        noneExpr: TextLiteral('None'),
      ).evaluate(implement),
      'Some(aa',
    );
  });
}

File: deno/commection/example_test.ts
import { assertEquals } from "jsr:@std/assert";
import {
  evaluate,
  hello,
  ifExpr,
  Implement,
  optionalMatch,
  textIsEmpty,
  textJoin,
  textLiteral,
} from "./example.ts";

const implement: Implement = {
  hello: () => Promise.resolve("ok!!!"),
};

Deno.test("text is empty", async () => {
  assertEquals(await evaluate(implement, textIsEmpty(textLiteral(""))), true);
});

Deno.test("text join", async () => {
  assertEquals(
    await evaluate(implement, textJoin(textLiteral("sample: "), hello)),
    "sample: ok!!!"
  );
});

Deno.test("if true", async () => {
  assertEquals(
    await evaluate(
      implement,
      ifExpr({
        condition: textIsEmpty(textLiteral("")),
        thenExpr: textLiteral("T"),
        elseExpr: textLiteral("F"),
      })
    ),
    "T"
  );
});

Deno.test("if false", async () => {
  assertEquals(
    await evaluate(
      implement,
      ifExpr({
        condition: textIsEmpty(textLiteral("aa")),
        thenExpr: textLiteral("T"),
        elseExpr: textLiteral("F"),
      })
    ),
    "F"
  );
});

Deno.test("optional match", async () => {
  assertEquals(
    await evaluate(
      implement,
      optionalMatch<string, string>({
        optional: textLiteral("aa"),
        some: (value) => textJoin(textLiteral("Some("), value),
        none: textLiteral("None"),
      })
    ),
    "Some(aa"
  );
});
